kpi/filters.py

```diff
@@ -13,52 +17,109 @@ class KpiObjectPermissionsFilter(object):
 
     def filter_queryset(self, request, queryset, view):
         user = request.user
+        if user.is_superuser and view.action != 'list':
+            # For a list, we won't deluge the superuser with everyone else's
+            # stuff. This isn't a list, though, so return it all
+            return queryset
+        # Governs whether unsubscribed (but publicly discoverable) objects are
+        # included. Exclude them by default
+        all_public = bool(strtobool(
+            request.query_params.get('all_public', 'false').lower()))
+
         model_cls = queryset.model
         kwargs = {
             'app_label': model_cls._meta.app_label,
             'model_name': model_cls._meta.model_name,
         }
         permission = self.perm_format % kwargs
 
-        strict_query= view.action == 'list'
-        return get_objects_for_user(user, permission, queryset, strict=strict_query)
+        if user.is_anonymous():
+            user = get_anonymous_user()
+            # Avoid giving anonymous users special treatment when viewing
+            # public objects
+            owned_and_explicitly_shared = queryset.none()
+        else:
+            owned_and_explicitly_shared = get_objects_for_user(
+                user, permission, queryset)
+        public = get_objects_for_user(
+            get_anonymous_user(), permission, queryset)
+        if view.action != 'list':
+            # Not a list, so discoverability doesn't matter
+            return owned_and_explicitly_shared | public
+
+        # For a list, do not include public objects unless they are also
+        # discoverable
+        try:
+            discoverable = public.filter(discoverable_when_public=True)
+        except FieldError:
+            try:
+                # The model does not have a discoverability setting, but maybe
+                # its parent does
+                discoverable = public.filter(
+                    parent__discoverable_when_public=True)
```

Comment:
Are we no longer doing nested collections (which would require traversing ancestors up to the root)?